File: CTCI_1-4_PalindromePermutation.js
/*

CTCI - 1.4 Palindrome Permutation
Given a string, write a function to check if it is a permutation of a palindrome.  
A palindrome is a word or phrase that is the same forwards and backwards.  
A permutation is a rearrangement of letters. The palindrome does not need to be limited to just dictionary words

Clarification Questions:
-- 

EXAMPLE:
-- Input: Tact Coa
-- Output: True (permutations: "taco cat", "atco cat", etc.)

(I) Input: String
(O) Output: Boolean
(C) Constraints: Optimize
(E) Edge Cases: Empty String, spaces in front / middle / end

Time Complexity: 
Space Complexity: 

Pseudocode:
-- Check length of str 
-- Create dict to store chars and counts
-- Check dict depending on length of str (odd = only 1 odd # char, even = even # char only)
*/

let palindromPermutation = (str) => {
  var chars = {};
  var even = false;

  var strLen = 0;
  for (var i = 0; i < str.length; i++) {
    if (str[i] != " ") {
      strLen++;
    }
  }

  if (strLen % 2 == 0) {
    even = true;
    // console.log("String is even");
  } else {
    // console.log("String is odd");
  }

  for (let i = 0; i < str.length; i++) {
    var char = str[i];
    if (char == " ") {
      continue;
    }
    if (!chars[char] && chars[char] != " ") {
      chars[char] = 1;
    } else if (chars[char] && chars[char] != " ") {
      chars[char]++;
    }
  }
  for (var num in chars) {
    if (even) {
      if (chars[num] % 2 != 0) {
        return false;
      }
    }
    return true;
    // console.log(num, chars[num]);
  }
  //   console.log(chars);
};

console.log(
  palindromPermutation("taco cat") === true,
  palindromPermutation("atco cat") === true,
  palindromPermutation(" rac  ecar rara ") === true,
  palindromPermutation("chirpingmermaid") === false, //failing
  palindromPermutation("aabbc") === true,
  palindromPermutation("aaaabbbbcc") === true,
  palindromPermutation("aabc") === false, //failing
  palindromPermutation("") === true //failing
);

File: CTCI_1-4_PalindromePermutation.py
# CTCI - 1.4 Palindrome Permutation
# Given a string, write a function to check if it is a permutation of a palindrome.
# A palindrome is a word or phrase that is the same forwards and backwards.
# A permutation is a rearrangement of letters. The palindrome does not need to be limited to just dictionary words

# Clarification Questions:
# -- What is a palindrome? Example?
# -- What is a permutation? Example?
# -- What is a palindrome permutation? Example?
# -- What characters do we count? Special chars? Capitol?

# EXAMPLE:
# -- Input: Tact Coa
# -- Output: True (permutations: "taco cat", "atco cat", etc.)

# (I) Input: String
# (O) Output: Boolean
# (C) Constraints: Optimize
# (E) Edge Cases: Empty String, spaces in front / middle / end, special chars, capitolization

# Time Complexity:
# Space Complexity:

# Pseudocode:
# -- Check length of str
# -- Create dict to store chars and counts
# -- Check dict depending on length of str (odd = only 1 odd # char, even = even # char only)

# Normalize - lowercase and get rid of strings


def isPalindromePerm(inputStr):
    inputStr = inputStr.replace(' ', "")
    inputStr = inputStr.lower()

    chars = dict()
    for letter in inputStr:
        if letter in chars:
            chars[letter] += 1
        else:
            chars[letter] = 1

    odd_count = 0
    for key, val in chars.items():
        if val % 2 != 0 and odd_count == 0:
            odd_count += 1
        elif val % 2 != 0 and odd_count != 0:
            return False
    return True


print(isPalindromePerm('Taco cAt') == True)
print(isPalindromePerm('Not a palin perm') == False)
